lib/maze/hooks/browser_hooks.rb

```diff
@@ -0,0 +1,38 @@
+# Contains logic for the Cucumber hooks when in Browser mode
+module Maze
+  class BrowserHooks
+    def after_configuration
+      config = Maze.config
+      if config.farm == :bs
+        # BrowserStack browser
+        tunnel_id = SecureRandom.uuid
+        config.capabilities = Maze::Capabilities.for_browser_stack_browser config.browser,
+                                                                           tunnel_id,
+                                                                           config.capabilities_option
+        Maze::BrowserStackUtils.start_local_tunnel config.bs_local,
+                                                   tunnel_id,
+                                                   config.access_key
+      end
+
+      # Create and start the relevant driver
+      if config.farm == :bs
+        selenium_url = "http://#{config.username}:#{config.access_key}@hub.browserstack.com/wd/hub"
+        Maze.driver = Maze::Driver::Browser.new :remote, selenium_url, config.capabilities
+      elsif config.farm == :local
+        Maze.driver = Maze::Driver::Browser.new Maze.config.browser.to_sym
+      end
+    end
+
+    def before
```

Comment:
To avoid having to always implement all of the hook methods, we could make an "abstract" class that noops and is extended by these hook classes. That way they only need to implement the methods that they require as they'll inherit the noops. i.e.

```rb
class Hooks
  def after_configuration
  end

  def before
  end

  def after
  end

  def at_exit
  end
end

class BrowserHooks < Hooks
  def after_configuration
    # stuff...
  end
end
```

Reply:
Ah, yes, that would be better.